### ISSUE TITLE
dlopen failed: cannot locate symbol "fgets_unlocked" referenced by "/data/app/com.zhinenggangqin-zz36yFPbZlpHeu83YEBT1w==/lib/arm/libglib-2.0.so"...

### ISSUE THREAD

**Issue by zhouzhihao07** [FluidSynth/fluidsynth#526]
### FluidSynth version 
2.0.4

McMidiPlayer* McMidiPlayer_new(const char* sf, double volumeGain)
{
    McMidiPlayer* me = (McMidiPlayer*)malloc(sizeof(McMidiPlayer));
    memset(me, 0, sizeof(McMidiPlayer));
    
    me->disableTrack = -1;
    me->settings = new_fluid_settings();
    //fluid_settings_setstr(me->settings, "audio.driver", "file");
    fluid_settings_setstr(me->settings, "audio.driver", "callback");
    fluid_settings_setint(me->settings, "audio.period-size", 1024);
    fluid_settings_setnum(me->settings, "synth.sample-rate", 8000);
    fluid_settings_setnum(me->settings, "synth.audio-channels", 2);
    fluid_settings_setnum(me->settings, "synth.reverb.room-size", 0.5);
    if(volumeGain > 0 && volumeGain <= 10){
        fluid_settings_setnum(me->settings, "synth.gain", volumeGain);  //音量增益
    }
    
    me->synth = new_fluid_synth(me->settings);

    int sfid = fluid_synth_sfload(me->synth, sf, 0);
    if(sfid == FLUID_FAILED){
        return NULL;
    }
    fluid_synth_program_select(me->synth, 0, sfid, 0, 0);

    me->adriver = new_fluid_audio_driver(me->settings, me->synth);

    me->midi_cur_msec = -1;
    me->midi_updated_timestamp = -1;
    
    me->userBPM = -1;
    return me;
}
![image](https://user-images.githubusercontent.com/8677018/55724525-2cf16300-5a3e-11e9-8353-81b043d740d0.png)

**Comment by atsushieno** [comment]
Most likely due to inconsistent set of libraries.

Possible cause: https://groups.google.com/forum/#!topic/android-ndk/Fu0b_Qpn0No

Possible fix: https://github.com/atsushieno/fluidsynth-midi-service-j/issues/12

**Comment by derselbst** [comment]
I understand this as a bug in the NDK which has been fixed in r17b. Closing.

**Comment by zhouzhihao07** [comment]
no  ndk 17b and
I  try r17-beta2 to compile，the same error.




------------------ 原始邮件 ------------------
发件人: "Tom M."<notifications@github.com>; 
发送时间: 2019年4月18日(星期四) 凌晨1:47
收件人: "FluidSynth/fluidsynth"<fluidsynth@noreply.github.com>; 
抄送: "残骸"<378623234@qq.com>; "Author"<author@noreply.github.com>; 
主题: Re: [FluidSynth/fluidsynth] dlopen failed: cannot locate symbol"fgets_unlocked" referenced by"/data/app/com.zhinenggangqin-zz36yFPbZlpHeu83YEBT1w==/lib/arm/libglib-2.0.so"...(#526)




Closed #526.

**Comment by atsushieno** [comment]
r17-beta2 (beta) is not the same as r17b (stable release, revision b).

**Comment by zhouzhihao07** [comment]
you look first picture,I can't find 17b.  only 17c,do you understand?


------------------ 原始邮件 ------------------
发件人: "Atsushi Eno"<notifications@github.com>;
发送时间: 2019年4月18日(星期四) 下午2:10
收件人: "FluidSynth/fluidsynth"<fluidsynth@noreply.github.com>;
抄送: "残骸"<378623234@qq.com>; "Author"<author@noreply.github.com>; 
主题: Re: [FluidSynth/fluidsynth] dlopen failed: cannot locate symbol"fgets_unlocked" referenced by"/data/app/com.zhinenggangqin-zz36yFPbZlpHeu83YEBT1w==/lib/arm/libglib-2.0.so"...(#526)




r17-beta2 (beta) is not the same as r17b (stable release, revision b).

**Comment by atsushieno** [comment]
You wrote

> I try r17-beta2 to compile，the same error.

This in English would mean "I tried r17-beta2 to compile it, but the same error occurred" - If you meant otherwise, explanation in full sensence would be great for understanding.

I assume you haven't tried r17c yet (no proof in the first screenshot for that), it's worth trying a *clean* build with r17c. (`git clean -xdf` and `git reset --hard` depending on whether you have local changes / development files or not)

**Comment by zhouzhihao07** [comment]
I have tried all the recent NDK versions, and 17c is the first version I used. I only tried the latest versions because of the fgets_unlock error of 17c. There will be problems with the fgets_unlock error of 17b (beta2) and the unread_lock error of 19c.




------------------ 原始邮件 ------------------
发件人: "Atsushi Eno"<notifications@github.com>;
发送时间: 2019年4月18日(星期四) 下午3:21
收件人: "FluidSynth/fluidsynth"<fluidsynth@noreply.github.com>;
抄送: "残骸"<378623234@qq.com>; "Author"<author@noreply.github.com>; 
主题: Re: [FluidSynth/fluidsynth] dlopen failed: cannot locate symbol"fgets_unlocked" referenced by"/data/app/com.zhinenggangqin-zz36yFPbZlpHeu83YEBT1w==/lib/arm/libglib-2.0.so"...(#526)




You wrote
  
I try r17-beta2 to compile，the same error.
  
This in English would mean "I tried r17-beta2 to compile it, but the same error occurred" - If you meant otherwise, explanation in full sensence would be great for understanding.
 
I assume you haven't tried r17c yet (no proof in the first screenshot for that), it's worth trying a clean build with r17c. (git clean -xdf and git reset --hard depending on whether you have local changes / development files or not)

**Comment by zhouzhihao07** [comment]
Have you ever really tried to compile it and use it, tell me where the ndk-17b is




------------------ 原始邮件 ------------------
发件人: "Atsushi Eno"<notifications@github.com>;
发送时间: 2019年4月18日(星期四) 下午3:21
收件人: "FluidSynth/fluidsynth"<fluidsynth@noreply.github.com>;
抄送: "残骸"<378623234@qq.com>; "Author"<author@noreply.github.com>; 
主题: Re: [FluidSynth/fluidsynth] dlopen failed: cannot locate symbol"fgets_unlocked" referenced by"/data/app/com.zhinenggangqin-zz36yFPbZlpHeu83YEBT1w==/lib/arm/libglib-2.0.so"...(#526)




You wrote
  
I try r17-beta2 to compile，the same error.
  
This in English would mean "I tried r17-beta2 to compile it, but the same error occurred" - If you meant otherwise, explanation in full sensence would be great for understanding.
 
I assume you haven't tried r17c yet (no proof in the first screenshot for that), it's worth trying a clean build with r17c. (git clean -xdf and git reset --hard depending on whether you have local changes / development files or not)

**Comment by atsushieno** [comment]
Not sure if you really understand what I wrote, but let me rephrase:

- r17b is NOT r17 beta 2
- it's worth trying a *clean* build with r17c.

> Have you ever really tried to compile it and use it, tell me where the ndk-17b is

You can find old NDK r17c (which is newer than r17b and still stable) from here: https://developer.android.com/ndk/downloads/older_releases.html

I am the latest Android build contributor and I keep building libfluidsynth. See my first comment that has a link to my set of prebuilt Android binaries. If you just want to use libfluidsynth on Android you don't even have to build it.

And it would help others understand to describe how you are building it. There are various ways to build it. (You didn't mention Android so it's been just a guess by the issue title.)

**Comment by zhouzhihao07** [comment]
I  know  r17b is NOT r17 beta 2

I try  a  build with r17c at the first time。It's 



I need to change some code, so I'm going to recompile with the your wiki



------------------ 原始邮件 ------------------
发件人: "Atsushi Eno"<notifications@github.com>;
发送时间: 2019年4月18日(星期四) 下午5:17
收件人: "FluidSynth/fluidsynth"<fluidsynth@noreply.github.com>;
抄送: "残骸"<378623234@qq.com>; "Author"<author@noreply.github.com>; 
主题: Re: [FluidSynth/fluidsynth] dlopen failed: cannot locate symbol"fgets_unlocked" referenced by"/data/app/com.zhinenggangqin-zz36yFPbZlpHeu83YEBT1w==/lib/arm/libglib-2.0.so"...(#526)




Not sure if you really understand what I wrote, but let me rephrase:
  
r17b is NOT r17 beta 2
 
it's worth trying a clean build with r17c.
   
Have you ever really tried to compile it and use it, tell me where the ndk-17b is
  
You can find old NDK r17c (which is newer than r17b and still stable) from here: https://developer.android.com/ndk/downloads/older_releases.html
 
I am the latest Android build contributor and I keep building libfluidsynth. See my first comment that has a link to my set of prebuilt Android binaries. If you just want to use libfluidsynth on Android you don't even have to build it.
 
And it would help others understand to describe how you are building it. There are various ways to build it. (You didn't mention Android so it's been just a guess by the issue title.)

**Comment by santoxyz** [comment]
i hit this issue compiling with ndk r21d and API 28.
Maybe it's related to the fact my phone has API 27 (Android 8)

how to deal with reccomandation in fluid-android-prepare.sh 
# Set Android target API level
# when compiling with clang use at least 28 as this makes sure that android provides the posix_spawn functions, so the compilation of gettext will (should) work out of the box
# it's probably a bug of gettext, if posix_spawn is not available it replaces it with its own implementation. Autotools of gettext set HAVE_POSIX_SPAWN==0 (which is correct) but for some reason REPLACE_POSIX_SPAWN==0 (which is wrong, as it should be 1).
export ANDROID_API=28

?

**Comment by atsushieno** [comment]
If you want to build with clang, you most likely have to build all dependencies with clang, not just Fluidsynth itself. gettext dependency mess is handled there. It perhaps needs fairly up to date revision of Cerbero build system.

**Comment by santoxyz** [comment]
Hi, thank you for support.
I adapted fluid-android-prepare.sh and fluid-android-compile.sh updating libraries versions.
using ndk21d, ANDROID_API=23 seems fine! 
(For now I only produced armv7a libs)
 How can I share my modified scripts? they seems to be out of tree gists.
https://github.com/FluidSynth/fluidsynth/wiki/BuildingForAndroid

**Comment by atsushieno** [comment]
I guess I was weighing in wherever I am not involved; that is an old build script that only cares one single ABI, different from doc/android. Maybe you can give @derselbst your diffs (either here or on those gists) ?

**Comment by derselbst** [comment]
The scripts are provided for educational purposes and for those, who want full manual control of the build process. So, they are intended for modifications. I think would be best to leave a comment on the gist where you share your experience.

**Comment by santoxyz** [comment]
Hi, here you can find my updated scripts
https://gist.github.com/santoxyz
